foreshadow/core/fs_logging.py

```diff
@@ -0,0 +1,264 @@
+"""Logging methods interfacing with a potential GUI system."""
+import atexit as _atexit
+import logging as _logging
+import sys as _sys
+import threading as _threading
+
+
+_logger = None
+_logger_lock = _threading.Lock()
+
+_this_module = _sys.modules[__name__]
+
+LEVELS = {
+    "debug": _logging.DEBUG,
+    "info": _logging.INFO,
+    "warning": _logging.WARNING,
+    "error": _logging.ERROR,
+    "critical": _logging.CRITICAL,
+}
+
+
+def get_logger():
+    """Return Foreshadow logger instance.
+
+    Will create and setup if needed, else will return the previously setup
+    logger.
+
+    Returns:
+        foreshadow logger instance.
+
+    """
+    global _logger
+
+    # if already created, don't acquire lock
+    if _logger:
+        return _logger
+
+    _logger_lock.acquire()
+
+    try:
+        if _logger is not None:
+            return _logger
+
+        # Get scoped Foreshadow logger.
+        _my_logger = _logging.getLogger("foreshadow")
+
+        _interactive = False
+        try:
+            # Only defined in interactive shells.
+            _interactive = True if _sys.ps1 else _interactive
+        except AttributeError:
+            # check python -i
+            _interactive = _sys.flags.interactive
+
+        if _interactive:
+            _my_logger.setLevel(LEVELS["info"])
+        else:
+            _my_logger.setLevel(LEVELS["warning"])
+        _stream_target = _sys.stderr
+
+        # Add Stream Handler based on if interactive or not.
+        _handler = _logging.StreamHandler(_stream_target)
+        _handler.setFormatter(_logging.Formatter(_logging.BASIC_FORMAT, None))
+        _my_logger.addHandler(_handler)
+
+        _logger = _my_logger
+        return _logger
+
+    finally:
+        _logger_lock.release()
+
+
+def _get_log_fn(level):
+    """Get the logging function for a particular method from this module.
+
+    Used internally in this script as the module will not yet have the
+    methods set.
+
+    Args:
+        level: The corresponding log function level. E.g. 'debug'
+
+    Returns:
+        The log function callable.
+
+    Raises:
+        ValueError: if logging level does not have a logging function.
+
+    """
+    log_fn = getattr(_this_module, level, None)
+    if log_fn is None:
+        raise ValueError("level: '{}' is not a valid logging level.")
+    return log_fn
+
+
+class SyncWrite(object):
+    """Single-instance object to handle buffered writing to file."""
+
+    # TODO turn this into synchronous Queue based writing.
+    def __init__(
+        self, buffer_size=100, outfile="gui_data.txt", overwrite=True
+    ):
+        super(SyncWrite, self).__init__()
+        self.threads = []
+        self._buffer_size = buffer_size
```

Comment:
If you change this to self.buffer_size, you get to use the validation you already wrote on input.

Reply:
right. will do.